### PR TITLE
fix(common/web): mock deadkey handling after rules manipulating context 📴

### PR DESCRIPTION
**Pull request by jahorton**
Happened upon this one while working on #7343.

The bug being addressed: a keyboard rule triggered manipulation of prior context that resulted in adding or removing a character before the caret, deadkey positions weren't being updated.  Subsequent rules would fail to match any deadkeys still in the context after the caret.

The scenario is quite niche, so it's unlikely to have had any major effect - especially since we're talking about post-caret deadkeys.  If any effect resulted from this, it'd have affected our predictive text quality.

@keymanapp-test-bot skip

This bug was discovered by our unit tests once I corrected a reference to point to `Mock` testing instead of `TouchAlias` testing, and the fix fully corrects it without any change to the unit test.  Unit tests are sufficient here.

### CODE REVIEW

File: common/web/keyboard-processor/src/text/outputTarget.ts [hunk 1]
```diff
@@ -439,12 +439,14 @@ namespace com.keyman.text {
...
           dn = this.caretIndex;
         }
+        this.adjustDeadkeys(-dn);
         this.text = this.text.kmwSubstr(0, this.caretIndex - dn) + this.getTextAfterCaret();
         this.caretIndex -= dn;
       }
     }
 
     insertTextBeforeCaret(s: string): void {
+      this.adjustDeadkeys(s._kmwLength());
```
Comment (jahorton) [inline common/web/keyboard-processor/src/text/outputTarget.ts:449]: Compare:

https://github.com/keymanapp/keyman/blob/1a7ad09fdb3e14a6d117e8555775745bece555d8/web/source/dom/targets/touchAlias.ts#L53-L67

**Comment by jahorton** [comment]
Hmm... there are a few notable issues with element focus management cropping up in the unit tests.  I'll need to investigate exactly what's causing the issues.

The gist I'm seeing so far:  whenever KMW decides it needs to re-focus on the should-remain-activeElement, sometimes that very act actually _unsets_ the activeElement - and that's breaking the affected unit tests.

**Comment by mcdurdin** [comment]
> Subsequent rules would fail to match any deadkeys still in the context after the caret.

How can a rule match a deadkey in the context _after_ the caret? Rules always match pre-caret. Furthermore, a deadkey that was after the caret is effectively gone, because caret movement (using mouse, cursor keys, etc) triggers an invalidation of context which drops all deadkey state. I'm worried that there are some incorrect assumptions about how deadkeys work?

File: web/source/dom/domEventHandlers.ts [hunk 2]
```diff
@@ -675,9 +675,7 @@ namespace com.keyman.dom {
       // This works OK for iOS, but may need something else for other platforms
       var elem: HTMLElement;
 
-      if(('relatedTarget' in e) && e.relatedTarget) {
-        elem = e.relatedTarget as HTMLElement;
-      }
+      elem = DOMEventHandlers.states._lastActiveElement;
```
Comment (mcdurdin) [inline web/source/dom/domEventHandlers.ts:678]: Is this an unrelated fix? If so, I think it should be in a separate PR for tracking and history.

**Review by mcdurdin — CHANGES_REQUESTED** [review]
I assume the unit test fixes are actually cleanup for other issues? It might be good to split them out into separate PRs.

The code change to domEventHandlers.ts, however, seems quite unrelated to the deadkey handling. Could we track that separately please?

**Comment by mcdurdin** [comment]
I've been thinking about this a bit more. This is a change to keystroke processing. Generally, I think it is super important that we do user testing (and if possible automated tests also) on any change to the keystroke processing code. So I think we need to:

1. provide an example of how it was going wrong (e.g. with a specific keyboard by comparison with Keyman 14 for Windows, and/or by comparison with KeymanWeb 10*)
2. do a user test on the change.


* Keyman 14 for Windows because it used the old Windows-core which was effectively our reference implementation. Keyman 15 for Windows is very close to that but still ironing out a few edge cases.
* KeymanWeb 10 because that was effectively our ref impl for web, and an old "known-good" version. Could alternatively use 2.x series. Not that it was 100% the same as Windows, nor necessarily 100% correct, but at least we need to demonstrate where this was diverging with Mocks.

File: web/source/dom/domEventHandlers.ts (hunk 2 above)
Comment (jahorton) [reply web/source/dom/domEventHandlers.ts:678]: This may have crept in from #7343; I'll double-check.  Work on that one is what led to discovery of this issue, after all.

File: web/tools/recorder/browserDriver.ts [hunk 3]
```diff
@@ -52,7 +52,7 @@ namespace KMWRecorder {
       // To be safe, we replicate the MouseEvent similarly to the keystroke event.
       var downEvent;
       var upEvent;
-      if(target['base'] && target instanceof HTMLDivElement) {
+      if((window['keyman'] as any).util.device.touchable) {
```
Comment (jahorton) [inline web/tools/recorder/browserDriver.ts:55]: This crept in from #7343; will fix.

**Comment by jahorton** [comment]
> I've been thinking about this a bit more. This is a change to keystroke processing.

Uh, how exactly?  This is a bug fix, not a change.  The implementation of the `Mock` type (within outputTarget.ts) was incomplete in this regard, mismatching its close parallel with touch-alias element handling.

**Comment by jahorton** [comment]
... And I merged in #7346 before I realized it was based on this PR, not `master`.  Oh well.

**Comment by mcdurdin** [comment]
> This is a bug fix, not a change

A bug fix is a change. A change in the right direction, certainly, but still a change.

**Comment by jahorton** [comment]
> > Subsequent rules would fail to match any deadkeys still in the context after the caret.
> 
> How can a rule match a deadkey in the context _after_ the caret? Rules always match pre-caret. Furthermore, a deadkey that was after the caret is effectively gone, because caret movement (using mouse, cursor keys, etc) triggers an invalidation of context which drops all deadkey state. I'm worried that there are some incorrect assumptions about how deadkeys work?

This particular function has existed since before KMW 10 and is the reason the unit test exists in the first place:

KMW 14+:

https://github.com/keymanapp/keyman/blob/497a8613db42fee2f812e4a5d48b11651bbd9eae/common/web/keyboard-processor/src/text/deadkeys.ts#L137-L154

KMW 10:

https://github.com/keymanapp/keyman/blob/f16872a99896a2fc208c68ac92a827820ab07e3f/web/source/kmwcallback.ts#L1250-L1263

Admittedly, I've always been a bit fuzzy on exactly what sort of case this was meant to address; the most straightforward way to double-check that its functionality is maintained uninterrupted is via a unit test that operates in violation of your intuition above.  If "it can even handle that case", we're in a good spot.

Therefore:

> 1. provide an example of how it was going wrong (e.g. with a specific keyboard by comparison with Keyman 14 for Windows, and/or by comparison with KeymanWeb 10*)

If you can provide me with a keyboard that motivated the existence of that code above... _then_ I can fulfill your request.

**Comment by jahorton** [comment]
> 1. provide an example of how it was going wrong (e.g. with a specific keyboard by comparison with Keyman 14 for Windows, and/or by comparison with KeymanWeb 10*)

A minor followup on this point:  due to the specific nature of how bugs related to this could surface, this request is literally impossible with regard to KMW 10.  KMW 10 did not support predictive text.  This bug is on Mocks, an input-receiving type that didn't even exist until KMW 12, and is currently triggered in production only by predictive-text-prep operations.

**Comment by jahorton** [comment]
> I assume the unit test fixes are actually cleanup for other issues? It might be good to split them out into separate PRs.

Well, about that.

This PR itself is cleanup for that other issue:  See #7343.  The "unit test fix" is that the test was pointing at a type we're about to deprecate in that PR, rather than the type it was always supposed to be testing.

Said unit-test fails once corrected without this PR's changes.  My investigation into why it failed... is why this PR exists.

**Comment by mcdurdin** [comment]
> Admittedly, I've always been a bit fuzzy on exactly what sort of case this was meant to address.

It's a long time ago, but I *think* it relates to IE's text selection support. I have a feeling it may be completely obsolete now. IE's selection management did not give an offset relative to start of text, so we had everything calculated as relative offsets to the 'selection' (caret if selection is empty). Hence, any changes to context required a recalculation of our cached deadkey offsets. However, once we have deadkey offsets in absolute positions, then deletion of content just requires deletion of deadkeys that fall to the right of the caret -- and that should already be working, of course.

We should do some careful testing, I guess, to see if this feeling of mine is correct, but if so, time for a little happy dance: simplifying complex logic is a Good Thing.

**Comment by jahorton** [comment]
> > Admittedly, I've always been a bit fuzzy on exactly what sort of case this was meant to address.
> 
> It's a long time ago, but I _think_ it relates to IE's text selection support. I have a feeling it may be completely obsolete now. IE's selection management did not give an offset relative to start of text, so we had everything calculated as relative offsets to the 'selection' (caret if selection is empty). Hence, any changes to context required a recalculation of our cached deadkey offsets. However, once we have deadkey offsets in absolute positions, then deletion of content just requires deletion of deadkeys that fall to the right of the caret -- and that should already be working, of course.
> 
> We should do some careful testing, I guess, to see if this feeling of mine is correct, but if so, time for a little happy dance: simplifying complex logic is a Good Thing.

Sounds like something that should be a separate PR, though - that involves far more substantial changes than what this PR brings.

I'm also a bit curious if we should maintain the relative offset bit anyway for things like #5258.

**Comment by mcdurdin** [comment]
> Well, about that.
> 
> This PR itself is cleanup for that other issue: see #7343. The "unit test fix" is that the test was pointing at a type we're about to deprecate in that PR, rather than the type it was always supposed to be testing.
> 
> Said unit-test fails once corrected without this PR's changes. My investigation into why it failed... is why this PR exists.

🤯

So confused... can you explain again? I don't understand why you are making changes against master that relate to a completely separate PR's changes? Why aren't those changes in #7343 or at least in a PR targeting #7343?

File: common/web/keyboard-processor/src/text/outputTarget.ts (hunk 1 above)
Comment (mcdurdin) [reply common/web/keyboard-processor/src/text/outputTarget.ts:449]: Why `s._kmwLength()` as opposed to `s.kmwLength()`?

**Comment by mcdurdin** [comment]
> I'm also a bit curious if we should maintain the relative offset bit anyway for things like #5258.

I'm okay with removing it entirely and then revisiting one day in full for that issue, rather than attempting to build on the currently shaky foundation for relative offsets.

**Comment by jahorton** [comment]
> So confused... can you explain again? I don't understand why you are making changes against master that relate to a completely separate PR's changes? Why aren't those changes in #7343  [...]

image

It already has quite the changeset.  This is orthogonal enough to split off as a separate thing.

_Technically_, what this addresses actually been a hidden bug for quite a while, as the unit test wasn't pointing at the right test target and allowed the related unit tests to pass when they should have failed.  

> or at least in a PR targeting #7343?

Because it's actually orthogonal enough to #7343 to implement first without any ill effects whatsoever.  And not only that - even if we decided to change our minds tomorrow and drop all work on #7343, I believe this is its own bug that should still be fixed.  Granted, there's the strong chance for a further-simplifying followup as noted in some of your comments above.

**Review by mcdurdin — APPROVED** [review]
OK, LGTM I think. I think the original problem I had here when reviewing was confusion to the other fixes that had crept in with the actual fix, which is in two parts: 1) the unit test was never actually testing Mocks, and 2) adjustDeadkeys wasn't called.

I'm going to call out for hopeful removal of adjustDeadkeys (and any correlated IE-specific adjustment code, if it exists) as a separate issue.

File: common/web/keyboard-processor/src/text/outputTarget.ts (hunk 1 above)
Comment (jahorton) [reply common/web/keyboard-processor/src/text/outputTarget.ts:449]: Most of the calls in that file use `_kmwLength`, so it follows the same pattern.

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 16.0.71-alpha](https://keyman.com/downloads/releases/alpha/16.0.71)
